libs/theseus_ffi/src/lib.rs

```diff
@@ -0,0 +1,308 @@
+        }
+    }
+}
+
+impl<T, E> From<Result<T, E>> for FfiResult<T, E> {
+    fn from(value: Result<T, E>) -> Self {
+        match value {
+            Ok(t) => FfiResult::Ok(t),
+            Err(e) => FfiResult::Err(e),
+        }
+    }
+}
+
+impl<T, E> Try for FfiResult<T, E> {
+    type Output = T;
+
+    type Residual = FfiResult<Infallible, E>;
+
+    fn from_output(output: Self::Output) -> Self {
+        Self::Ok(output)
+    }
+
+    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
+        match self {
+            Self::Ok(v) => ControlFlow::Continue(v),
+            Self::Err(e) => ControlFlow::Break(FfiResult::Err(e)),
+        }
+    }
+}
+
+impl<T, E> FromResidual for FfiResult<T, E> {
+    fn from_residual(residual: <Self as core::ops::Try>::Residual) -> Self {
+        match residual {
+            FfiResult::Err(e) => Self::Err(e),
+        }
+    }
+}
+
+#[derive(Debug)]
+#[repr(C)]
+pub enum FfiOption<T> {
+    Some(T),
+    None,
+}
+
+impl<T> From<FfiOption<T>> for Option<T> {
+    fn from(value: FfiOption<T>) -> Self {
+        match value {
+            FfiOption::Some(t) => Some(t),
+            FfiOption::None => None,
+        }
+    }
+}
+
+impl<T> From<Option<T>> for FfiOption<T> {
+    fn from(value: Option<T>) -> Self {
+        match value {
+            Some(t) => FfiOption::Some(t),
+            None => FfiOption::None,
+        }
+    }
+}
+
+impl<T> Try for FfiOption<T> {
+    type Output = T;
+
+    type Residual = FfiOption<Infallible>;
+
+    fn from_output(output: Self::Output) -> Self {
+        Self::Some(output)
+    }
+
+    fn branch(self) -> ControlFlow<Self::Residual, Self::Output> {
+        match self {
+            Self::Some(v) => ControlFlow::Continue(v),
+            Self::None => ControlFlow::Break(FfiOption::None),
+        }
+    }
+}
+
+impl<T> FromResidual for FfiOption<T> {
+    fn from_residual(residual: <Self as core::ops::Try>::Residual) -> Self {
+        match residual {
+            FfiOption::None => Self::None,
+        }
+    }
+}
+
+#[derive(Debug, Clone, Copy)]
+#[repr(C)]
+pub enum Error {
+    NotFound,
+    PermissionDenied,
+    ConnectionRefused,
+    ConnectionReset,
+    HostUnreachable,
+    NetworkUnreachable,
+    ConnectionAborted,
+    NotConnected,
+    AddrInUse,
+    AddrNotAvailable,
+    NetworkDown,
+    BrokenPipe,
+    AlreadyExists,
+    WouldBlock,
+    NotADirectory,
+    IsADirectory,
+    DirectoryNotEmpty,
+    ReadOnlyFilesystem,
+    FilesystemLoop,
+    StaleNetworkFileHandle,
+    InvalidInput,
+    InvalidData,
+    TimedOut,
+    WriteZero,
+    StorageFull,
+    NotSeekable,
+    FilesystemQuotaExceeded,
+    FileTooLarge,
+    ResourceBusy,
+    ExecutableFileBusy,
+    Deadlock,
+    CrossesDevices,
+    TooManyLinks,
+    InvalidFilename,
+    ArgumentListTooLong,
+    Interrupted,
+    Unsupported,
+    UnexpectedEof,
+    OutOfMemory,
+    Other,
+}
+
+#[derive(Debug, Clone)]
+#[repr(C)]
+pub struct FatPointer {
```

Comment:
this can impl Send/Sync if and only if it actually comes from types that impl Send/Sync. Here I don't see that being the case yet, as it's always transmuted to/from something. 

Instead, it'd be cleaner and safer to use a From/Into impl on limited types that you know are safe to convert to/from fat pointers like this. Given a set of closed conversion methods like that, _then_ it might be okay to impl Send/Safe.